compiler/rustc_borrowck/src/diagnostics/move_errors.rs

```diff
@@ -623,30 +623,51 @@ impl<'infcx, 'tcx> MirBorrowckCtxt<'_, 'infcx, 'tcx> {
-                    // we point at the type on the parameter from the definition of the outer
-                    // function:
-                    //
-                    // error[E0507]: cannot move out of `foo`, a captured
-                    //               variable in an `Fn` closure
-                    //   --> file.rs:14:25
-                    //    |
-                    // 13 | fn do_stuff(foo: Option<Foo>) {
-                    //    |             ---  ----------- move occurs because `foo` has type
-                    //    |             |                `Option<Foo>`, which does not implement
-                    //    |             |                the `Copy` trait
-                    //    |             captured outer variable
-                    // 14 |     require_fn_trait(|| async {
-                    //    |                      -- ^^^^^ `foo` is moved here
-                    //    |                      |
-                    //    |                      captured by this `Fn` closure
-                    // 15 |         if foo.map_or(false, |f| f.foo()) {
-                    //    |            --- variable moved due to use in coroutine
-                    use_span = param.ty_span;
+                    let upvar = &self.upvars[upvar_field.index()];
+                    let upvar_hir_id = upvar.get_root_variable();
+                    use_span = match self.infcx.tcx.parent_hir_node(upvar_hir_id) {
+                        hir::Node::Param(param) => {
+                            // Instead of pointing at the path where we access the value within a
+                            // closure, we point at the type on the parameter from the definition
+                            // of the outer function:
+                            //
+                            // error[E0507]: cannot move out of `foo`, a captured
+                            //               variable in an `Fn` closure
+                            //   --> file.rs:14:25
+                            //    |
+                            // 13 | fn do_stuff(foo: Option<Foo>) {
+                            //    |             ---  ----------- move occurs because `foo` has type
+                            //    |             |                `Option<Foo>`, which does not
+                            //    |             |                implement the `Copy` trait
+                            //    |             captured outer variable
+                            // 14 |     require_fn_trait(|| async {
+                            //    |                      -- ^^^^^ `foo` is moved here
+                            //    |                      |
+                            //    |                      captured by this `Fn` closure
+                            // 15 |         if foo.map_or(false, |f| f.foo()) {
+                            //    |            --- variable moved due to use in coroutine
```

Comment:
These large comments make this function harder to read imo. I would personally prefer a link to the affected test case instead